src/sqlfluff/dialects/dialect_tsql.py

```diff
@@ -5140,3 +5143,66 @@ class CreateExternalDataSourceStatementSegment(BaseSegment):
             ),
         ),
     )
+
+
+class OpenJsonWithClauseSegment(BaseSegment):
+    """A `WITH` clause of an `OPENJSON()` table-valued function.
+
+    https://learn.microsoft.com/en-us/sql/t-sql/functions/openjson-transact-sql?view=sql-server-ver16#with_clause
+    """
+
+    type = "openjson_with_clause"
+
+    match_grammar = Sequence(
+        "WITH",
+        Bracketed(
+            Delimited(
+                AnyNumberOf(
+                    Sequence(
+                        Ref("ColumnReferenceSegment"),
+                        Ref("DatatypeSegment"),
+                        Sequence(
+                            Ref("QuotedLiteralSegment"),  # column_path
+                            optional=True,
+                        ),
+                        Sequence(
+                            "AS",
+                            "JSON",
+                            optional=True,
+                        ),
+                    ),
+                    min_times=1,
+                ),
+            ),
+        ),
+    )
+
+
+class OpenJsonSegment(BaseSegment):
+    """An `OPENJSON()` table-valued function.
+
+    https://learn.microsoft.com/en-us/sql/t-sql/functions/openjson-transact-sql?view=sql-server-ver16#syntax
+    """
+
+    type = "openjson_segment"
+
+    match_grammar = Sequence(
+        "OPENJSON",
+        Bracketed(
+            Delimited(
+                OneOf(
+                    Ref("QuotedLiteralSegmentOptWithN"),  # jsonExpression
+                    Ref("ColumnReferenceSegment"),
+                    Ref("ParameterNameSegment"),
+                ),
+                Sequence(
+                    Ref("QuotedLiteralSegment"),  # path
+                    optional=True,
```

Comment:
this optional flag isn't needed, as the sequence is a positional arg in delimited, which extends OneOf()